lib/web/desktop.go

```diff
@@ -173,48 +174,175 @@ func proxyWebsocketConn(ws *websocket.Conn, con net.Conn) error {
 	return trace.NewAggregate(retErrs...)
 }
 
+const actionToggle = "toggle"
+
+// playbackAction is a message passed from the playback client
+// to the server over the websocket connection in order to modify
+// the playback state.
+type playbackAction struct {
+	// Action is one of actionToggle | TODO: actionMove
+	// actionToggle toggles the playbackState.playing
+	Action string `json:"action"`
+}
+
+// playbackState is a thread-safe struct for managing
+// the global state of the playback websocket connection.
+type playbackState struct {
+	ws      *websocket.Conn
+	playing bool // false means the playback is paused
+	wsOpen  bool
+	mu      sync.RWMutex
+}
+
+// GetPlaying returns ps.playing
+func (ps *playbackState) GetPlaying() bool {
+	ps.mu.RLock()
+	defer ps.mu.RUnlock()
+	return ps.playing
+}
+
+// TogglePlaying toggles the boolean state of ps.playing
+func (ps *playbackState) TogglePlaying() {
+	ps.mu.Lock()
+	defer ps.mu.Unlock()
+	ps.playing = !ps.playing
+}
+
+// GetWsOpen gets the wsOpen value. It should be called at the top of continuously
+// looping goroutines that use playbackState, who should return if it returns false.
+func (ps *playbackState) GetWsOpen() bool {
+	ps.mu.RLock()
+	defer ps.mu.RUnlock()
+	return ps.wsOpen
+}
+
+// Close closes the websocket connection and sets wsOpen to false.
+// It should be deferred by all the goroutines that use playbackState,
+// in order to ensure that when one goroutine closes, all the others do too
+// (when used in conjunction with GetWsOpen).
+func (ps *playbackState) Close() {
+	ps.ws.Close()
```

Comment:
Is it okay to close the websocket connection multiple times? Do we see any errors or anything if this happens?

Reply:
If you `Close` a websocket that's already closed it returns a `net.ErrClosed` (this implementation was just ignoring any `Close` errors). I've added error handling and wrapped this with a `sync.Once`.